src/pages/instances/Events.tsx

```diff
@@ -7,19 +7,46 @@ import { useQueryClient } from "@tanstack/react-query";
 import { queryKeys } from "util/queryKeys";
 import { useOperations } from "context/operationsProvider";
 import { useNotify } from "@canonical/react-components";
+import useEventListener from "util/useEventListener";
 
 const EVENT_HANDLER_DELAY = 250;
 const WS_RETRY_DELAY_MULTIPLIER = 250;
 const MAX_WS_CONNECTION_RETRIES = 5;
+const EVENT_WS_STALE_TIMEOUT_MS = 3600_000;
 
 const Events: FC = () => {
   const { isAuthenticated } = useAuth();
   const eventQueue = useEventQueue();
   const queryClient = useQueryClient();
   const notify = useNotify();
   const [eventWs, setEventWs] = useState<WebSocket | null>(null);
+  const [reconnectWsId, setReconnectWsId] = useState(0);
+  const [wsOpenTime, setWsOpenTime] = useState(0);
   const { refetchOperations } = useOperations();
 
+  const getCurrentTime = () => {
+    return new Date().getTime();
+  };
+
+  const reconnectWsOnFocusTab = () => {
+    if (document.visibilityState !== "visible") {
+      return;
+    }
+    if (!eventWs) {
+      setReconnectWsId((prevId) => prevId + 1);
+      return;
+    }
+    // If the ws was open for more than EVENT_WS_STALE_TIMEOUT_MS, close it and reconnect.
+    const timeSinceOpen = getCurrentTime() - wsOpenTime;
+    const isStale = timeSinceOpen > EVENT_WS_STALE_TIMEOUT_MS;
+    if (isStale && eventWs) {
+      eventWs.close();
+      setEventWs(null);
+      setReconnectWsId((prevId) => prevId + 1);
```

Comment:
Looks like doing both of these would trigger the effect below, that depends on `[eventWs, isAuthenticated, reconnectWsId]`, twice. Maybe we can just do `setEventWs(null)`? And this makes me wonder: do we actually need the `reconnectWsId` state? Can't we just use the `eventWs` and toggle it between null and ws?
In the check above, where we do:
```
    if (!eventWs) {
      setReconnectWsId((prevId) => prevId + 1);
      return;
    }
```
we could instead directly call the function that is currently inlined with the `useEffect`.

Reply:
Actually, we don't even need to call `setEventWs(null)` here. Because we do `eventWs.close();` and the close handler will already do `setEventWs(null);`.

Reply:
> we could instead directly call the function that is currently inlined with the useEffect.

That won't work, we need the cleanup of the use effect.

> Actually, we don't even need to call setEventWs(null) here. Because we do eventWs.close(); and the close handler will already do setEventWs(null);.

True and the close call is guaranteed to trigger either onclose or onerror. Will simplify this bit.

Reply:
The use effect calls the returned function on unmount. We can't do that with calling `myEffectHandler` directly. The returned function is ignored.

Reply:
useEffect is not calling the function directly. **<ins>It's passing the function as a parameter instead of inlining it</ins>**. From my comment above:
```
useEffect(myEffectHandler, [eventWs, isAuthenticated]);
```
Also quoting again from the comment above:
> Alternatively, keep the effect handler inline and just change the check here:
```
    // Force reconnect if the ws is not open.
    if (!eventWs) {
      connectEventWs();
      return;
    }
```
> And I still think the reconnectWsId state is superfluous.

Reply:
Correct, it is called directly only in the `reconnectWsOnFocusTab` context and you don't need the return value there, so you just ignore it there and all is good. Or what's the harm of having a return value that is not used?

Anyway, I think we should focus on my second idea, which looks cleaner/better, doesn't touch the inlined useEffect, and still doesn't require the additional state.

Reply:
> Correct, it is called directly only in the reconnectWsOnFocusTab context and you don't need the return value there, so you just ignore it there and all is good. Or what's the harm of having a return value that is not used?

We need the callback to be called once the component unmounts to close the websocket. That is why the use effect is in place and needed.

> Anyway, I think we should focus on my second idea, which looks cleaner/better, doesn't touch the inlined useEffect, and still doesn't require the additional state.

That would leave us with multiple connected websockets. We shouldn't consume more than one at a time.

Reply:
> We need the callback to be called once the component unmounts to close the websocket. That is why the use effect is in place and needed.

Answered already: https://github.com/canonical/lxd-ui/pull/1328#discussion_r2166536850

> That would leave us with multiple connected websockets. We shouldn't consume more than one at a time.

```
    if (!eventWs) {
      connectEventWs();
      return;
    }
```
We call `connectEventWs` only when `eventWs` is null. So how can we have multiple connected websockets?

Reply:
Moving to https://github.com/canonical/lxd-ui/pull/1328#discussion_r2166903507, resolving this.